source/fuzz/fuzzer_pass_add_composite_inserts.cpp

```diff
@@ -0,0 +1,164 @@
+                });
+
+        // If there are no available constants, then return.
+        if (available_composite_constants.empty()) {
+          return;
+        }
+
+        // Choose randomly one available composite constant.
+        auto available_composite_constant =
+            available_composite_constants[GetFuzzerContext()->RandomIndex(
+                available_composite_constants)];
+
+        // Take a random component of the chosen composite constant.
+        auto num_constants = available_composite_constant->NumInOperands();
+        uint32_t index_to_replace =
+            GetFuzzerContext()->GetRandomIndexForAccessChain(num_constants);
+        auto constant_to_replace_id =
+            available_composite_constant->GetSingleWordInOperand(
+                index_to_replace);
+        auto constant_to_replace_type_id = GetIRContext()
+                                               ->get_def_use_mgr()
+                                               ->GetDef(constant_to_replace_id)
+                                               ->type_id();
+
+        // Look for available constants that have the same type id as the
+        // |constant_to_replace|.
+        std::vector<opt::Instruction*> available_constants =
+            FindAvailableInstructions(
+                function, block, instruction_iterator,
+                [this, instruction_descriptor, constant_to_replace_type_id](
+                    opt::IRContext* ir_context,
+                    opt::Instruction* instruction) -> bool {
+
+                  if (instruction->type_id() != constant_to_replace_type_id) {
+                    return false;
+                  }
+
+                  return fuzzerutil::IdIsAvailableBeforeInstruction(
+                      ir_context,
+                      FindInstruction(instruction_descriptor, ir_context),
+                      instruction->result_id());
+                });
+
+        // Choose randomly one available constant.
+        auto available_constant =
+            available_constants[GetFuzzerContext()->RandomIndex(
+                available_constants)];
+
+        auto new_result_id = GetFuzzerContext()->GetFreshId();
+
+        // Insert an OpCompositeInsert instruction which copies
+        // |available_composite_constant| and in the copied composite constant
+        // replaces |constant_to_replace| with |available_constant|.
+        FindInstruction(instruction_descriptor, GetIRContext())
+            ->InsertBefore(MakeUnique<opt::Instruction>(
+                GetIRContext(), SpvOpCompositeInsert,
+                available_composite_constant->type_id(), new_result_id,
+                opt::Instruction::OperandList(
+                    {{SPV_OPERAND_TYPE_ID, {available_constant->result_id()}},
+                     {SPV_OPERAND_TYPE_ID,
+                      {available_composite_constant->result_id()}},
+                     {SPV_OPERAND_TYPE_LITERAL_INTEGER, {index_to_replace}}})));
```

Comment:
Fuzzer passes must not directly change the module. They can only change the module via transformations. The output from spirv-fuzz is a list of transformations that change the module. We also output the final modified module, but the most important output is the list of transformations. We need to be able to replay the list of transformations and replay a subset of the list of transformations.